#PBS -N 04_Recruit_ShortReads
#PBS -l nodes=1:ppn=2
#PBS -l mem=2gb
#PBS -l walltime=24:00:00
#PBS -q microcluster #iw-shared-6
#PBS -o 00_log/04_Recruit_ShortReads/${sID}.out
#PBS -e 00_log/04_Recruit_ShortReads/${sID}.err

cd $PBS_O_WORKDIR

query_array=(${Query_Path}/*.fa) # adjust file extension to metagenomes
query=${query_array[$PBS_ARRAYID]}
qID=`basename $query | cut -d. -f1`
out=${Output_Dir}/${qID}_${sID}.blast

echo 'query =' $query
echo 'Ref Fasta =' $Ref_Fasta
echo 'qID =' $qID
echo 'sID =' $sID
echo 'out =' $out

magicblast -query $query -db $Ref_Fasta -num_threads 10 -no_unaligned -splice F -outfmt tabular -parse_deflines T -out $out

python $filter -i $out -pid 0.95 -pml 0.9 -rl 70 -o ${out}.fltrd

if [ -s ${out}.fltrd ]
  then
	python $collect -b ${out}.fltrd -q $query -o ${out}.fltrd.fasta
fi

if [ -s ${out}.fltrd.fasta ]
  then
	rm ${out} ${out}.fltrd
fi

# Other Flags available for Magic Blast:
#  -perc_identity <Real, 0..100>
#  -score <String> Cutoff score for accepting alignments. Can be expressed as a number or a function of read length: L,b,a for a * length + b. Default = `20'
#  -paired Input query sequences are paired
#  -no_query_id_trim Do not trim '.1', '/1', '.2', or '/2' at the end of read ids for SAM format and paired runs
#  -no_discordant Suppress discordant alignments for paired reads

